graph/pull_v2.go

```diff
@@ -250,66 +250,72 @@ func (p *v2Puller) pullV2Tag(tag, taggedName string) (verified bool, err error)
 			logrus.Debugf("error getting image v1 json: %v", err)
 			return false, err
 		}
-		downloads[i].img = img
-		downloads[i].digest = manifest.FSLayers[i].BlobSum
-
 		p.graph.Retain(p.sessionID, img.ID)
 		layerIDs = append(layerIDs, img.ID)
 
 		// Check if exists
 		if p.graph.Exists(img.ID) {
 			logrus.Debugf("Image already exists: %s", img.ID)
+			out.Write(p.sf.FormatProgress(stringid.TruncateID(img.ID), "Already exists", nil))
 			continue
 		}
-
 		out.Write(p.sf.FormatProgress(stringid.TruncateID(img.ID), "Pulling fs layer", nil))
 
-		downloads[i].err = make(chan error)
-		downloads[i].out = pipeWriter
-		go p.download(&downloads[i])
+		d := &downloadInfo{
+			img:    img,
+			digest: manifest.FSLayers[i].BlobSum,
+			err:    make(chan error, 1),
```

Comment:
I see.

So the operation proceeds on the buffered, allowing the goroutine to proceed. Perhaps, the issue is that we don't guarantee that the channel is consumed. It might be better to have a control channel that can be paired in a select statement. If the operation is cancelled, we close the control channel and the operation on the error channel doesn't _have_ to proceed.

It might be acceptable in the course of this PR to add a TODO identifying that the the buffer channel represents a larger problem with this function and we need to address it accordingly.